### ISSUE TITLE
Explore GPU training

### ISSUE THREAD

**Issue by sergia-ch** [tournesol-app/tournesol#134]
**TLDR: on a Tesla P100 with Database from May 2021, got 33% speedup in time (compared to tournesol server). Can we make it go faster?**

- On Colab instance with the current database (May 20th 2021)
```
(pid=10887) 2021-05-20 18:03:35.216477: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1720] Found device 0 with properties:
(pid=10887) pciBusID: 0000:00:04.0 name: Tesla P100-PCIE-16GB computeCapability: 6.0
(pid=10887) coreClock: 1.3285GHz coreCount: 56 deviceMemorySize: 15.90GiB deviceMemoryBandwidth: 681.88GiB/s
```

- Training itself takes 11 minutes instead of 15 (30% speedup), but the whole process takes 13.4 mins (instead of 20mins, 33% speedup) with ~10% GPU load and ~400MB on GPU
- To run with a GPU: add `n_gpus: 1` to `.gin` config `tune` resources (that's it)
- Additionally, set `export TF_FORCE_GPU_ALLOW_GROWTH=true` to use only part of GPU
- Memory usage on CPU is higher

![image](https://user-images.githubusercontent.com/1012270/119025827-0e3b5580-b994-11eb-9665-38acd7b1ce6b.png)

![image](https://user-images.githubusercontent.com/1012270/119024926-2363b480-b993-11eb-84af-d0644610c538.png)

Why is it so slow? Can it be made faster?

Related: tournesol-app/tournesol-meta#2

**Comment by Zelgunn** [comment]
The time needed to fetch the data and send it to the GPU, relative to the time needed for a single train step, is very high.
When the data is fully located on the GPU, we can obtain a significant speed-up (~x50, 20sec for convergence in my experiment)

Here is the code I developed for this experiment:
https://gist.github.com/Zelgunn/5a617802eac2c81ee04941b798be950f

**Comment by lenhoanglnh** [comment]
Another strategy is decentralized learning. When contributors learn locally their individual scores (given global scores), and send them to the server, which updates global scores accordingly. The details of how best to do this (when, and if we need iterating these updates) is a research question.

Learning for contributors with a small number of ratings could then perhaps be done in their browser (maybe for <1k ratings).
Eventually, we should think of an external application for other contributors, and for contributors that may want to use algorithms able to learn...
